### PR TITLE
Replace O(N) lookup with O(1) for task outputs (master)

### LINKED ISSUE #3255: Speed up dependency checks for large numbers of outputs
We ran into a case where suites were taking a very long time to validate.  This was covered in #1776 and follow on tickets, but didn't quite solve the problem.  We tracked the behavior down to an O(N**2) operation in the number of task outputs.

Steps to reproduce: Have a very large number of task outputs (~1000)? in the graph that will be loaded.

Our solution was to replace a list with a set in the taskdef initialization that allowed checking for membership to be done in O(1) time instead of O(N) time (PR to follow)

### PR DESCRIPTION
**Pull request by trwhitcomb**
(see #3259 for 7.8.x PR and discussion)

This speeds up suite validation significantly for tasks with
large numbers of outputs.  The list used required an O(N)
check for membership which, when combined with being run for
each output resulted in O(N**2) complexity.

These changes close #3255

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Appropriate tests are included (unit and/or functional).

- [x] Appropriate change log entry included.

- [x] No documentation update required.

### CODE REVIEW

**Comment by trwhitcomb** [comment]
Sorry for the messier history - I forgot to fetch from "upstream" before pushing.  Also changed names on this and #3259 to be consistent with other dual-version PRs.

**Comment by hjoliver** [comment]
> Sorry for the messier history - I forgot to fetch from "upstream" before pushing. Also changed names on this and #3259 to be consistent with other dual-version PRs.

You can clean up the branch history (i.e. "rewrite history") with `git rebase -i`.

File: cylc/flow/config.py [hunk 1]
```diff
@@ -1617,7 +1617,7 @@ def generate_taskdefs(self, orig_expr, left_nodes, right, seq):
             # Record custom message outputs.
             for item in self.cfg['runtime'][name]['outputs'].items():
                 if item not in taskdef.outputs:
-                    taskdef.outputs.append(item)
+                    taskdef.outputs.add(item)
```
Comment (hjoliver) [inline cylc/flow/config.py]: @trwhitcomb - with reference to https://github.com/cylc/cylc-flow/pull/3259#discussion_r310849186 ... you could ditch the set membership test at 1619 and just add the items (that would definitely be more "Pythonic" whether or not there's a marginal speed advantage either way.).

**Comment by matthewrmshin** [comment]
Hi @trwhitcomb Any update on this one? Really like to see this merged.

**Comment by hjoliver** [comment]
@matthewrmshin - if @trwhitcomb has gone AWOL one of us can tweak the branch as suggested above.

**Comment by matthewrmshin** [comment]
@trwhitcomb I have decided to take over this one so we have a good chance of getting this for the alpha release next week. I hope you don't mind. (The commits will still be attributed to you in `git log`.)

**Comment by matthewrmshin** [comment]
Superseded by #3351.

**Pull request closed without merging**